xmodule/docs/decisions/0003-library-content-block-schema.rst

```diff
@@ -0,0 +1,201 @@
+
+Evolving the library_content block schema
+#########################################
+
+Status
+******
+
+**Provisional**
+
+Subject to change due to implementation learnings and stakeholder feedback.
+
+Context
+*******
+
+The library_content block is an existing block type which allows users to include a specified number of blocks into a course, randomly picked per-learner from a "V1" (modulestore-backed) content library, optionally filtered by the blocks' ``capa_type`` (only applicable to ``problem`` blocks). We are incrementally improving this feature:
+
+* We are soon adding support for "V2" (blockstore-backed) content libraries to the library_content block. This will enable an improved library authoring experience and assist our transition away from modulestore and MongoDB. We need this not to break existing usages of the library_content block, and we need there to be a period of simultaneous V1/V2 support overlap to enable a reversible migration of libraries from modulestore to blockstore. Eventually, V1 support will be removed.
+
+* Next, we want to add support for non-randomized, hand-picked reference of library blocks. This is is a highly requested feature across the community.
+
+* In the future, we would like to support other modes: in particular, we want authors to eventually be able to hand-pick a set of blocks from a library, and *then* have the LMS randomly select a specified number of blokcs for each learner.
+
+We need to choose how to evolve the schema of the library_content block to support all of these incremental improvements. This is the current schema:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Field
+     - Scope, Type
+     - Default
+     - Example
+     - Description
+
+   * - source_library_id
+     - settings, str|None
+     - None
+     - "library-v1:OEX+Lib1"
+     - Key of V1 source library; None means unselected.
+
+   * - source_library\_ version
+     - settings, str|None
+     - None
+     - TBD
+     - Version of V1 source library (a MongoID string). None means latest.
+
+   * - mode
+     - settings, str
+     - "random"
+     - "random"
+     - How blocks should be selected. Only valid value is "random".
+
+   * - capa_type
+     - settings, str
+     - "any"
+     - "symbolicresponse"
+     - Unless "any": Include only problems, and only problems of this response type.
+
+   * - children
+     - special, list of XBlocks
+     - []
+     - ``block-v1:...+type@problem+block@A``
+       ``block-v1:...+type@problem+block@B``
+       ``block-v1:...+type@problem+block@D``
+       ``block-v1:...+type@problem+block@F``
+     - Copy of library contents, filtered by capa_type. Course author may apply edits that are local to this library_content block usage. Edits to child blocks' settings persist across library version upgrades. Edits to child blocks' content are wiped out when library version is upgraded.
```

Comment:
I was confused by the presence of this `children` field on the list and didn't understand it from this description. Then I realized it's just the standard `children` field that all XBlocks with children have.

I would suggest something like description in case anyone else is lead astray like I was: `Standard XBlock field that holds the IDs of the child XBlocks. This field is only populated after the library content block is saved, at which time the matching XBlocks (e.g. filtered by capa_type) from the library are copied to become children of this block. Learners will often only see a subset of these children (up to max_count). Course author may apply edits that are local to these copied child blocks. Such edits to child blocks' Scope.settings fields will persist across library version upgrades, whereas edits to child blocks' Scope.content fields are wiped out when library version is upgraded.`